Fix issue #382 (multiple publish output files with same relative path)

LucaMaccarini opened the pull request:
Fixes #382: On Windows, runtime files from backend packages are currently flattened during copy. This means that all files from subfolders end up in the same output directory, causing conflicts when multiple files share the same name.

## Proposed fix

1. Correct the file [`LLamaSharpBackend.props`](https://github.com/LucaMaccarini/LLamaSharp/blob/fix-issue-%23382/LLama/runtimes/build/LLamaSharpBackend.props) as shown in the changes included in this PR.

2. When building for Windows, and after installing the Windows-specific backend packages, make sure to add `ExcludeAssets="Native"` to the `PackageReference` in the `.csproj`, as shown below:

```xml
<PackageReference Include="LLamaSharp.Backend.Cpu" Version="0.24.0" ExcludeAssets="Native"/>
```

Although `ExcludeAssets="Native"` is strictly necessary only for Windows backends when building on Windows, it shouldn’t cause issues for other backends. To be safe, it can be applied to all imported backend packages. I have tested this with both `LLamaSharp.Backend.Cpu` and `LLamaSharp.Backend.Android` installed in the same MAUI project. This attribute tells MSBuild not to copy the runtime files from the NuGet package to the output directory, because copying is already handled correctly in the `LLamaSharpBackend.props` file updated in step 1.

## How to Apply this fix before it is merged

If you urgently need to fix this issue, you can temporarily apply this change by copying the corrected `LLamaSharpBackend.props` content (from step 1) into each backend package directory, e.g.:

```
C:\Users\<your user>\.nuget\packages\<llamasharp backend>\<version>\build\netstandard2.0\<llamasharp backend>.props
```

and then import the package as indicated in step 2. 

**Note:** this modification will be lost if the package is updated.

## Future improvements:
Avoid using the default `Runtimes` folder, since MSBuild always flattens its contents building for windows causing conflicts. Using a custom folder for runtime files would eliminate the need to set `ExcludeAssets="Native"` in the package reference, and then change the `LLamaSharpBackend.props` file to correctly fetch runtime files from the custom folder.

## Tests
I tested this fix on a limited scenario with a MAUI project having both `LLamaSharp.Backend.Cpu` and `LLamaSharp.Backend.Android` installed. The goal was to ensure the app works on both types of devices. Given the large heterogeneity of the backends, I hope this fix will work for other backends as well.

martindevans commented:
Thanks for looking into this. #382 has been a long standing issue that a few people (including me) have had a look at but failed to solve, I'm happy to see the fix turns out to be so simple!

I'll find some time to test this out soon, I'd really like to get it into the next update if possible.

Would you be interested in developing a followup PR with your suggested improvements? Removing the extra installation step and making it just work would be fantastic.

m0nsky commented:
Hi @LucaMaccarini , does this PR need any additional changes? I'm not sure what you mean by that last note (have you found out if we need to start with `runtimes/` or `../` ?). I'm currently running into the same issue on linux (but have not yet tested this PR).

LucaMaccarini commented:
Hi @m0nsky, Thanks a lot for testing and for the detailed feedback

I actually suspected that the path part should have remained `runtimes/deps/`.
In my local setup, I was using `nuget.exe` directly, passing the `.nuspec` file to create the package. Since the `.nuspec` files are located one level deeper respect to `deps` folder, I assumed I needed to prepend `../deps`.

Example of the command I used locally:

```
.\nuget.exe pack "C:\Users\luca.maccarini\Desktop\luca\myProjects\LLamaSharp\LLamaSharp\LLama\runtimes\build\LLamaSharp.Backend.Cuda12.Linux.nuspec" -Version 0.25.0
```

I’ll fix the initial path in this PR so that it uses `runtimes/deps/`.

**Could you kindly share how you usually generate the NuGet packages on your side? (what script or command do you run). That would help me align with your workflow and avoid mismatches like this.**

Regarding the point you mentioned about the `LLamaSharp.Backend.Cuda12.Linux` package also including the CPU backend: this is expected, because in its `.nuspec` there is an explicit dependency on `LLamaSharp.Backend.Cpu` (introduced by @MartinEvans in commit `02eedd94`).
Here’s the relevant metadata:

```xml
<dependencies>
    <dependency id="LLamaSharp.Backend.Cpu" version="$version$" />
</dependencies>
```

That dependency is the reason why the CPU backend is included when packing the CUDA12 backend.

m0nsky commented:
Yes, absolutely, here are my steps:

- Download `nuget.exe` and add to PATH (environment variable, Windows 11)
- Clone your WIP branch, build the project
- Pack the nuget packages, using my method described [here](https://gist.github.com/m0nsky/8d0f56618838326a07850ff441b821f0)

Basically, I put that `nuget_pack.bat` inside the `LLamaSharp` folder, and run it. After packing the nuget packages, the `.nupkg` files are located in `LLamaSharp/temp`. 

What I meant by the CPU/CUDA12 issue, is that the CUDA12 runtimes seem to be completely missing when building my application. 

When I extract `LLamaSharp.Backend.Cuda12.Linux.3.0.1.nupkg` using 7zip, I can see that `LLamaSharpRuntimes\linux-x64\native\cuda12` exists in the nuget package. However, when building my application which has the `LLamaSharp` and `LLamaSharp.Backend.Cuda12` nuget packages installed (from the local source, which we just built), the runtimes folder only contains the CPU backends.

LSXAxeller commented:
I am attempting to use a locally built version of the LLamaSharp CPU and Vulkan backends to test a specific fix, but I'm encountering a build error during the publish step.

Here is my process:

1.  I cloned the `fix-issue-#382` branch from your fork: [https://github.com/LucaMaccarini/LLamaSharp/tree/fix-issue-%23382](https://github.com/LucaMaccarini/LLamaSharp/tree/fix-issue-%23382).
2.  I used the `nuget_pack.bat` script to build and pack the local packages.
3.  I copied the generated packages to a `local-packages` directory in my repository's root.
4.  I added the following `nuget.config` file to my repository root to define the local package source for the CPU and Vulkan backends:

```xml
<?xml version="1.0" encoding="utf-8"?>
<configuration>
    <packageSources>
        <clear />
        <!-- Define local source using a relative path -->
        <add key="local-packages" value="./local-packages" />
        <!-- Add the official nuget.org source -->
        <add key="nuget.org" value="https://api.nuget.org/v3/index.json" protocolVersion="3" />
    </packageSources>
    <packageSourceMapping>
        <packageSource key="nuget.org">
            <package pattern="*" />
        </packageSource>
        <packageSource key="local-packages">
            <package pattern="LLamaSharp.Backend.Cpu" />
            <package pattern="LLamaSharp.Backend.Vulkan" />
            <package pattern="LLamaSharp.Backend.Vulkan.Linux" />
            <package pattern="LLamaSharp.Backend.Vulkan.Windows" />
        </packageSource>
    </packageSourceMapping>
</configuration>
```

In my IDE's NuGet Package Manager, I selected the `local-packages` source and installed the new CPU and Vulkan backend packages. The base `LlamaSharp` package is still from the official nuget.org feed.

When I try to publish my application with the following command, the build fails:

`dotnet publish -c Release -r win-x64 --self-contained -f net8.0-windows`

I receive the `NETSDK1152` error, which a conflict due to multiple publish output files having the same relative path.

Here is the error output:

```
E:\Others\Native\Workspace\ProseFlow\ProseFlow.UI>dotnet publish -c Release -r win-x64 --self-contained -f net8.0-windows
Determining projects to restore...
All projects are up-to-date for restore.
ProseFlow.Core -> E:\Others\Native\Workspace\ProseFlow\ProseFlow.Core\bin\Release\net8.0\ProseFlow.Core.dll
ProseFlow.Application -> E:\Others\Native\Workspace\ProseFlow\ProseFlow.Application\bin\Release\net8.0\ProseFlow.Application.dll
ProseFlow.Infrastructure -> E:\Others\Native\Workspace\ProseFlow\ProseFlow.Infrastructure\bin\Release\net8.0\ProseFlow.Infrastructure.dll
ProseFlow.UI -> E:\Others\Native\Workspace\ProseFlow\ProseFlow.UI\bin\Release\net8.0-windows\win-x64\ProseFlow.dll

C:\Program Files\dotnet\sdk\8.0.414\Sdks\Microsoft.NET.Sdk\targets\Microsoft.NET.ConflictResolution.targets(112,5): error NETSDK1152: Found multiple publish output files with the same relative path:

C:\Users\PRIME\.nuget\packages\llamasharp.backend.cpu\0.25.0\runtimes\win-x64\native\avx\ggml-base.dll,
C:\Users\PRIME\.nuget\packages\llamasharp.backend.cpu\0.25.0\runtimes\win-x64\native\avx2\ggml-base.dll,
C:\Users\PRIME\.nuget\packages\llamasharp.backend.cpu\0.25.0\runtimes\win-x64\native\avx512\ggml-base.dll,
C:\Users\PRIME\.nuget\packages\llamasharp.backend.cpu\0.25.0\runtimes\win-x64\native\noavx\ggml-base.dll,
C:\Users\PRIME\.nuget\packages\llamasharp.backend.cpu\0.25.0\runtimes\win-x64\native\avx\ggml-cpu.dll,
C:\Users\PRIME\.nuget\packages\llamasharp.backend.cpu\0.25.0\runtimes\win-x64\native\avx2\ggml-cpu.dll,
C:\Users\PRIME\.nuget\packages\llamasharp.backend.cpu\0.25.0\runtimes\win-x64\native\avx512\ggml-cpu.dll,
C:\Users\PRIME\.nuget\packages\llamasharp.backend.cpu\0.25.0\runtimes\win-x64\native\noavx\ggml-cpu.dll,
C:\Users\PRIME\.nuget\packages\llamasharp.backend.cpu\0.25.0\runtimes\win-x64\native\avx\ggml.dll,
C:\Users\PRIME\.nuget\packages\llamasharp.backend.cpu\0.25.0\runtimes\win-x64\native\avx2\ggml.dll,
C:\Users\PRIME\.nuget\packages\llamasharp.backend.cpu\0.25.0\runtimes\win-x64\native\avx512\ggml.dll,
C:\Users\PRIME\.nuget\packages\llamasharp.backend.cpu\0.25.0\runtimes\win-x64\native\noavx\ggml.dll,
C:\Users\PRIME\.nuget\packages\llamasharp.backend.cpu\0.25.0\runtimes\win-x64\native\avx\llama.dll,
C:\Users\PRIME\.nuget\packages\llamasharp.backend.cpu\0.25.0\runtimes\win-x64\native\avx2\llama.dll,
C:\Users\PRIME\.nuget\packages\llamasharp.backend.cpu\0.25.0\runtimes\win-x64\native\avx512\llama.dll,
C:\Users\PRIME\.nuget\packages\llamasharp.backend.cpu\0.25.0\runtimes\win-x64\native\noavx\llama.dll,
C:\Users\PRIME\.nuget\packages\llamasharp.backend.cpu\0.25.0\runtimes\win-x64\native\avx\mtmd.dll,
C:\Users\PRIME\.nuget\packages\llamasharp.backend.cpu\0.25.0\runtimes\win-x64\native\avx2\mtmd.dll,
C:\Users\PRIME\.nuget\packages\llamasharp.backend.cpu\0.25.0\runtimes\win-x64\native\avx512\mtmd.dll,
C:\Users\PRIME\.nuget\packages\llamasharp.backend.cpu\0.25.0\runtimes\win-x64\native\noavx\mtmd.dll.

[E:\Others\Native\Workspace\ProseFlow\ProseFlow.UI\ProseFlow.UI.csproj::TargetFramework=net8.0-windows]
```

Any guidance would be greatly appreciated.

LucaMaccarini commented:
Thank you both @m0nsky and @LSXAxeller for your feedback.
I believe I have fixed the problem, which was caused by the .props files for the Vulkan and CUDA12 packages having incorrect names. I have renamed them to match their package names, allowing MSBuild to import them correctly. This ensures that the runtime DLLs under LLamaSharpRuntimes\<OS>\native\vulkan and LLamaSharpRuntimes\<OS>\native\Cuda12 are copied to the build output.

m0nsky commented:
> Thank you both @m0nsky and @LSXAxeller for your feedback. I have fixed the problem, which was caused by the .props files for the Vulkan and CUDA12 packages having incorrect names. I have renamed them to match their package names, allowing MSBuild to import them correctly. This ensures that the runtime DLLs under LLamaSharpRuntimes<OS>\native\vulkan and LLamaSharpRuntimes<OS>\native\Cuda12 are copied to the build output.

Great job, I will test this tonight!

> However, it is still packing the Linux and macOS binaries into the `runtimes` folder, can we avoid that? Also, is there a way to specify the AVX version to use (e.g., by setting `DefineConstants` in the *.csproj file) to help MSBuild flatten the `runtimes` folder content directly into the application directory?
> 
> I can also confirm that adding any backend (Vulkan in my case) only adds the CPU backend binaries to the build, not the specific backend itself (e.g., Vulkan, CUDA).

I could understand not packing Linux/macOS when building for Windows only (if this is possible), however, I'm not sure if we should flatten the AVX levels, as I build my applications once and deploy them on different servers (AVX2 / AVX512) and rely on the automatic AVX level detection.

m0nsky commented:
Just tried the latest version, published to a linux server w/ CUDA, all working as expected.

LSXPrime commented:
While building my .NET application for macOS (target `osx-arm64`), I encountered the following warning:

```
>   dotnet publish ProseFlow.UI  -c Release  -r osx-arm64  --self-contained  -p:PublishSingleFile=true -p:UseAppHost=true -f net8.0-macos  -o "publish"

/Users/runner/.dotnet/packs/Microsoft.macOS.Sdk.net8.0_15.0/15.0.8319/targets/Xamarin.Shared.Sdk.targets(1897,3): warning :
The file '/Users/runner/.nuget/packages/llamasharp.backend.cpu/0.25.1/LLamaSharpRuntimes/osx-arm64/native/ggml-metal.metal' 
does not specify a 'PublishFolderType' metadata, and a default value could not be calculated. 
The file will not be copied to the app bundle. [/Users/runner/work/ProseFlow/ProseFlow/ProseFlow.UI/ProseFlow.UI.csproj::TargetFramework=net8.0-macos]
```

Since the build is running on GitHub Actions and I don't have a macOS machine to test, I'm not sure if this is a critical issue. Will the missing `ggml-metal.metal` file cause a runtime error or prevent the application from functioning correctly?

LSXPrime commented:
@LucaMaccarini I'm sorry I forgot to share an update—I've tackled this issue and have been busy handling several other issues caused by Apple macOS.

I fixed the issue by adding the condition `Condition="'$(RuntimeIdentifier)' == ''"` to your `ItemGroup`, so binaries are copied only when no RuntimeIdentifier is specified. If a RuntimeIdentifier is present, the corresponding platform-specific binaries are copied instead. This ensures the `runtimes` folder contains only the target platform's binaries.

LLamaSharpBackend.props
```xml
    <ItemGroup Condition="'$(RuntimeIdentifier)' != ''">
        <Content Include="$(MSBuildThisFileDirectory)..\..\LLamaSharpRuntimes\$(RuntimeIdentifier)\**">
            <CopyToOutputDirectory>PreserveNewest</CopyToOutputDirectory>
            <Visible>false</Visible>
            <Link>runtimes\$(RuntimeIdentifier)\%(RecursiveDir)%(Filename)%(Extension)</Link>
        </Content>
    </ItemGroup>

    <!-- Fallback for when RuntimeIdentifier is not specified -->
    <ItemGroup Condition="'$(RuntimeIdentifier)' == ''">
        <Content Include="$(MSBuildThisFileDirectory)..\..\LLamaSharpRuntimes\**\*.*">
            <CopyToOutputDirectory>PreserveNewest</CopyToOutputDirectory>
            <Visible>false</Visible>
            <Link>runtimes\%(RecursiveDir)%(Filename)%(Extension)</Link>
        </Content>
    </ItemGroup>
```

If you're using the publish command and specifying a RID (like win-x64 or osx-arm64), only the corresponding binaries should be copied to the runtimes folder. However, if your project references a backend like `LLamaSharp.Backend.Cpu` in a separate layer—such as an "Infrastructure" layer—then that setup can affect the publishing behavior, potentially including unnecessary or mismatched binaries.

In my case
> 
> ProseFlow/
> ├── ProseFlow.Core/
> ├── ProseFlow.Application/
> ├── ProseFlow.Infrastructure/ // LLamaSharp & Backends Layer (has `<PackageReference Include="LLamaSharp.Backend.Cpu" Version="0.25.1" />`)
> └── ProseFlow.UI/ // Main Layer (The one you build or publish like `dotnet publish ProseFlow.UI -c Release -r win-x64 --self-contained`)

In that case, you'll need to add the following to the layer's `*.csproj` file (e.g. `ProseFlow.Infrastructure.csproj` in my case) so that it can detect the `RuntimeIdentifier`—since the layer is built as a generic assembly, .NET only specifies the `RuntimeIdentifier` for the main layer.

```xml
    <PropertyGroup>
        <DetectedArch Condition="'$([System.Runtime.InteropServices.RuntimeInformation]::OSArchitecture)' == 'X64'">x64</DetectedArch>
        <DetectedArch Condition="'$([System.Runtime.InteropServices.RuntimeInformation]::OSArchitecture)' == 'Arm64'">arm64</DetectedArch>
    </PropertyGroup>

    <PropertyGroup>
        <RuntimeIdentifier Condition="$([MSBuild]::IsOSPlatform('Windows'))">win-x64</RuntimeIdentifier>
        <RuntimeIdentifier Condition="$([MSBuild]::IsOSPlatform('Linux'))">linux-x64</RuntimeIdentifier>
        <RuntimeIdentifier Condition="$([MSBuild]::IsOSPlatform('OSX')) AND '$(DetectedArch)' == 'arm64'">osx-arm64</RuntimeIdentifier>
        <RuntimeIdentifier Condition="$([MSBuild]::IsOSPlatform('OSX')) AND '$(DetectedArch)' == 'x64'">osx-x64</RuntimeIdentifier>
    </PropertyGroup>
```

If you'd like to test it quickly, I've uploaded a [prebuilt NuGet package here](https://github.com/LSXPrime/ProseFlow/blob/master/local-packages/LLamaSharp.Backend.Cpu.0.25.1.nupkg) to work with my GitHub build workflow

martindevans commented:
Thanks for all the work on this @LucaMaccarini and @m0nsky!
